### PR TITLE
block coinbase  spec 

### PR DESCRIPTION
**Pull request by DreamWuGit**
this PR depends on https://github.com/appliedzkp/zkevm-specs/pull/77 to add block coinbase op code spec with test.

### CODE REVIEW

**Comment by icemelon** [comment]
@lispc @HAOYUatHZ Could you help take a look at the spec PR?

**Comment by lispc** [comment]
LGTM

**Review by lispc — APPROVED** [review]

**Review by HAOYUatHZ — APPROVED** [review]
LGTM

File: src/zkevm_specs/evm/execution/block_coinbase.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+
+
+def coinbase(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.COINBASE)
+    address = instruction.stack_push()
+    # in real circuit check address raw data is 160 bit length (20 bytes)
```
Comment (HAOYUatHZ) [inline src/zkevm_specs/evm/execution/block_coinbase.py:9]: btw, why check in real circuit but not here?

File: src/zkevm_specs/evm/typing.py [hunk 2]
```diff
@@ -4,6 +4,34 @@
...
+    gas_limit: U64
+    block_number: U256
+    time: U256
+    difficulty: U256
+    base_fee: U256
+    previous_256_block_hashes: Sequence[U256]
+
+    def __init__(
+        self,
+        coinbase: U160 = 0x10,
```
Comment (HAOYUatHZ) [inline src/zkevm_specs/evm/typing.py:18]: any reference for the default coinbase address?

**Comment by HAOYUatHZ** [comment]
So I actually review it again. This PR in general looks good. just would like to learn about the answers to my 2 questions.

File: src/zkevm_specs/evm/typing.py (hunk 2 above)
Comment (DreamWuGit) [reply src/zkevm_specs/evm/typing.py:22]: currently block fields are filled into block context table, such table will be as pub input eventually.

File: src/zkevm_specs/evm/execution/block_coinbase.py (hunk 1 above)
Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/block_coinbase.py]: in python, U160 type already stands for 20 bytes, for now evm circuit also have type `Address` represents it limitation

File: src/zkevm_specs/evm/typing.py (hunk 2 above)
Comment (DreamWuGit) [reply src/zkevm_specs/evm/typing.py:22]: block fields will be public input in the future , so just pass them for now .

**Review by icemelon — COMMENTED** [review]
I have a question in general. There is no miner or difficulty in layer 2. So what should opcode `COINBASE` and `DIFFICULTY` return?

File: specs/opcode/41COINBASE.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+## Circuit behavior
+
+1. construct block context table 
+2. do bussmapping lookup for stack write operation
+3. other implicit check: bytes length
+
+## Constraints
+
+1. opId = 0x41
+2. state transition:\
```
[the remark this answers is not in the document]
Comment (DreamWuGit) [reply specs/opcode/41COINBASE.md]: seems pycharm add it automatically following by `:`,  have removed them.

**Comment by DreamWuGit** [comment]
> 
For no miner or difficulty condition, return zero as placeholder, as there are no more other constrains except stack bussmapping & lookup .

File: specs/opcode/41COINBASE.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+1. opId = 0x41
+2. state transition:  
+   gc + 1 ( 1 stack write)  
+   stack_pointer - 1  
+   pc + 1  
+   gas + 2
+3. lookups:  2  
+   `address` is on the top of stack  
+   `address` is in block context table
+
```
Comment (icemelon) [inline specs/opcode/41COINBASE.md:29]: remove extra empty line

File: specs/opcode/41COINBASE.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+## Exceptions
+
+1. gas out:  
+   remaining gas is not enough
+2. stack overflow:  
+   stack is full, stack pointer = 0
+
+## Code
+
+Refer to src/zkevm_specs/evm/execution/coinbase.py
```
Comment (icemelon) [inline specs/opcode/41COINBASE.md:42]:
```suggestion
Refer to `src/zkevm_specs/evm/execution/coinbase.py`
```

File: src/zkevm_specs/evm/execution/block_coinbase.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from ..instruction import Instruction, Transition
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/block_coinbase.py:1]: Just use `coinbase` as the file name?

Comment (DreamWuGit) [reply src/zkevm_specs/evm/execution/block_coinbase.py:1]: for `coinbase` case is ok, but I consider there are others block's op code (time, number etc), if use file name like  time.py, number.py , easily  cause ambiguous meaning compare to system files. 
So I prefer to keep prefix for all , like block_coinbase.py, block_time.py, block_number.py etc.

**Comment by icemelon** [comment]
close this one as a PR has been created to upstream

**Pull request closed without merging**